src/Stratis.Bitcoin.Features.Consensus/ProvenBlockHeaders/ProvenBlockHeaderStore.cs

```diff
@@ -0,0 +1,374 @@
+                        cachedHeaders.AddRange(repositoryHeaders);
+                }
+            }
+
+            return cachedHeaders;
+        }
+
+        /// <inheritdoc />
+        public async Task<HashHeightPair> CurrentTipHashHeightAsync()
+        {
+            this.CurrentTipHashHeight = await this.provenBlockHeaderRepository.GetTipHashHeightAsync().ConfigureAwait(false);
+
+            return this.CurrentTipHashHeight;
+        }
+
+        /// <inheritdoc />
+        public async Task<ProvenBlockHeader> GetTipAsync()
+        {
+            this.CurrentTipHashHeight = await this.CurrentTipHashHeightAsync();
+
+            return await this.GetAsync(this.CurrentTipHashHeight.Height).ConfigureAwait(false);
+        }
+
+        /// <inheritdoc />
+        public void AddToPending(ProvenBlockHeader provenBlockHeader, HashHeightPair newTip)
+        {
+            this.PendingCache.AddOrUpdate(newTip.Height, provenBlockHeader, (key, value) => { return provenBlockHeader; });
+
+            this.pendingTipHashHeight = newTip;
+        }
+
+        /// <summary>Saves pending <see cref="ProvenBlockHeader"/> items to the <see cref="ProvenBlockHeaderRepository"/>.</summary>
+        private async Task SaveAsync()
+        {
+            if (this.PendingCache.Count == 0)
+            {
+                this.logger.LogTrace("(-)[PROVEN_BLOCK_HEADER_CACHE_EMPTY]");
+                return;
+            }
+
+            using (new StopwatchDisposable(o => this.performanceCounter.AddInsertTime(o)))
+            {
+                using (await this.lockobj.LockAsync().ConfigureAwait(false))
+                {
+                    var pendingHeaders = new List<ProvenBlockHeader>();
+
+                    for (int i = this.CurrentTipHashHeight.Height; i <= this.pendingTipHashHeight.Height; i++)
+                    {
+                        ProvenBlockHeader cachedHeader = this.GetHeaderFromPendingCache(i);
+
+                        if (cachedHeader != null)
+                        {
+                            pendingHeaders.Add(cachedHeader);
+                            this.PendingCache.TryRemove(i, out ProvenBlockHeader removedItem);
+                        }
+                    }
+
+                    if (pendingHeaders.Count > 0)
+                        await this.provenBlockHeaderRepository.PutAsync(pendingHeaders, this.pendingTipHashHeight).ConfigureAwait(false);
+
+                    if (this.PendingCache.Count == 0)
+                        this.pendingTipHashHeight = null;
+                }
+            }           
+        }
+
+        /// <summary>Gets <see cref="ProvenBlockHeader"></see> items from pending cache with specific key if present.</summary>
+        /// <param name="key">Item's key.</param>
+        /// <returns><see cref="ProvenBlockHeader"></see> if cache contains the item, <c>null</c> otherwise.</returns>
+        private ProvenBlockHeader GetHeaderFromPendingCache(int key)
```

Comment:
This method is a bit useless

Reply:
Well leave i as is, its actually ok.

Reply:
Pleas change the names of the lists to be GetHeaderFromPending and GetHeaderFromCache (store is missleading in this case)

aLso rename `this.PendingCache` to `this.Pending`

Reply:
Spoke to @dangershony  keeping this.